### ISSUE TITLE
[aws-codebuild] Add support to grant SSM permissions to project 

### ISSUE THREAD

**Issue by BrianFarnhill** [aws/aws-cdk#11055]
With CodeBuild having support for SSM sessions, we should incorporate a property that allows these permissions to be included in the IAM role (in the same way as the permissions for build reports are done using the `grantReportGroupPermissions` property.

### Use Case

Using SSM to troubleshoot a build environment is a super useful new feature, but you need to add IAM permissions to the role of the build project to enable it to work. This can be done manually at the moment, but a cleaner property to enable it can let developers troubleshoot environments easier without having to remember the specific permissions to add.

### Proposed Solution

Add a property `grantSSMPermissions` to the Project construct, and have it add the below IAM policy to the project:

```
    buildProject.addToRolePolicy(new iam.PolicyStatement({
      actions: [
        "ssmmessages:CreateControlChannel",
        "ssmmessages:CreateDataChannel",
        "ssmmessages:OpenControlChannel",
        "ssmmessages:OpenDataChannel"
      ],
      effect: iam.Effect.ALLOW,
      resources: [ "*" ]
    }));
```

I'm not fussed about the default status of the property (I'm inclinded to think on by default as it's relatively harmless, but it is still adding a permission so it might not be ideal to make the assumption - given the grantReportGroupPermissions is default though this could be done the same way)

* [X] :wave: I may be able to implement this feature request (after re:Invent, I've got a lot on!) 
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by skinny85** [comment]
Hey @BrianFarnhill ,

can you elaborate on what integration with SSM is this helpful for? I'm not aware of such feature for CodeBuild, is this something new?

Thanks,
Adam

**Comment by BrianFarnhill** [comment]
Hey @skinny85 

It's for session manager, you can connect to a build environment to run commands interactively, super useful for troubleshooting issues you aren't able to replicate outside of the service.

https://aws.amazon.com/about-aws/whats-new/2020/07/aws-codebuild-now-supports-accessing-build-environments-with-aws-session-manager/

I'm building a demo of this in to my reinvent session. Having CDK support for the permission would make it easier for builders to start using the feature.

Does that help?

 - B

**Comment by skinny85** [comment]
Thanks @BrianFarnhill , that does help.

Reading the docs, I came across this:

> If you have Auditing and logging session activity enabled in Systems Manager preferences, the CodeBuild service role must also have additional permissions. The permissions are different, depending on where the logs are stored.

Does this make implementing this feature even more tricky?

**Comment by BrianFarnhill** [comment]
Ahhh that's fair actually, I had just been looking at enabling the ability to connect and run commands and not the logging. Let me have a think about it and I'll go back over the docs this week. Maybe there's a property here like this?

```
sessionManagerConfig:{
    enabled: true,
    loggingTarget: [something to flag that here]
}
```

Definitely need to think this through more than my initial suggestion, good call out!

**Comment by BrianFarnhill** [comment]
Looking through the docs then, I would think we could implement this in a property like this:

```
sessionManagerConfig: {
    allowConnections: boolean,
    s3LoggingTarget?: s3.IBucket,
    cloudwatchLoggingTarget?: logs.ILogGroup
}
```

If we use the allowConnections property to determine if the main permissions I initially suggested get added, and then optionally add permissions for S3 or CloudWatch based on the targets being provided as additional data. None of this would go on by default in this case, but it would be very straightforward to implement as a developer. If the bucket/log group are made in external templates you could use the corresponding `fromName` style methods to get the references to pass in. Do you think that could work?

**Comment by skinny85** [comment]
I think that's fine, but I would modify your first proposal slightly:

```ts
export interface SessionManagerConnection {
  // I don't like this enabled property... seems like we're forcing customers to tell us they want this twice.
  // If loggingTarget below is required, I would definitely get rid of this property
  readonly enabled: boolean;

  readonly loggingTarget?: LoggingTarget; // maybe this is required? not sure
}

export abstract class LoggingTarget {
  public static fromS3Bucket(bucket: s3.IBucket): LoggingTarget { ... }

  public static fromCloudWatchLogs(logGroup: logs.ILogGroup): LoggingTarget { ... }

  // ...
}
```

And you would use it like so:

```ts
new Project(this, 'Project', {
  // other properties...
  sessionManagerConnection: {
    enabled: true,
    loggingTarget: LoggingTarget.fromS3Bucket(myBucket),
  },
});
```

**Comment by BrianFarnhill** [comment]
I agree with your point about telling us what they want twice - the issue is though that the logging config is entirely optional. If you don't want to do any logging of session manager sessions, then you don't need to specify a logging config at all, so there needs to be a way to specify that you want the core permissions applied without either of the possible logging permissions being added to it, so the use cases would look like this:

### Session manager with no logging targets
```
new Project(this, 'Project', {
    // other properties...
    sessionManagerConnection: {
        enabled: true,
    },
});
```

### Session manager with an S3 logging target
```
new Project(this, 'Project', {
    // other properties...
    sessionManagerConnection: {
        enabled: true,
        loggingTarget: LoggingTarget.fromS3Bucket(myBucket),
    },
});
```

### Session manager with an CloudWatch logging target
```
new Project(this, 'Project', {
    // other properties...
    sessionManagerConnection: {
        enabled: true,
        loggingTarget: LoggingTarget.fromLogGroup(myLogGroup),
    },
});
```

What do you think of that approach?

**Comment by skinny85** [comment]
I don't hate it. One thing we could do if we really wanted to get rid of the `enabled` property, which kind of sucks, is turn this into a method on `Project`:

```ts
const project = new Project(this, 'Project', {
    // other properties...
});
project.enableConnectingFromSessionManager(LoggingTarget.fromLogGroup(myLogGroup));
// loggingTarget is an optional argument to the enableConnectingFromSessionManager() method
```

**Comment by BrianFarnhill** [comment]
@skinny85 I like that a lot more actually, with logging target being optional that gives us a nice neat way to do it. That's definitely how we should look at doing it.
